src/wallet/scriptpubkeyman.cpp

```diff
@@ -290,6 +290,47 @@ bool LegacyScriptPubKeyMan::GetReservedDestination(const OutputType type, bool i
     return true;
 }
 
+bool LegacyScriptPubKeyMan::TopUpInactiveHDChain(const CKeyID seed_id, int64_t index, bool internal)
+{
+    LOCK(cs_KeyStore);
+
+    if (m_storage.IsLocked()) return false;
+
+    auto it = m_inactive_hd_chains.find(seed_id);
+    if (it == m_inactive_hd_chains.end()) {
+        return false;
+    }
+
+    CHDChain& chain = it->second;
+
+    // Top up key pool
+    int64_t target_size = std::max(gArgs.GetArg("-keypool", DEFAULT_KEYPOOL_SIZE), (int64_t) 0);
+
+    // "size" of the keypools. Not really the size, actually the difference between index and the chain counter
+    // Since chain counter is 1 based and index is 0 based, one of them needs to be offset by 1.
+    int64_t kp_size = (internal ? chain.nInternalChainCounter : chain.nExternalChainCounter) - (index + 1);
+
+    // count number of missing keys (internal, external)
+    // make sure the keypool of external and internal keys fits the user-selected target (-keypool)
+    int64_t missing = std::max(std::max((int64_t) target_size, (int64_t) 1) - kp_size, (int64_t) 0);
+
+    WalletBatch batch(m_storage.GetDatabase());
+    for (int64_t i = missing; i > 0; --i) {
+        CPubKey pubkey(GenerateNewKey(batch, chain, internal));
+        AddKeypoolPubkeyWithDB(pubkey, internal, batch);
```

Comment:
This doesn't seem right. It adds the key to setInternalKeyPool or setExternalKeyPool, but I don't think we want to do that?

Reply:
Removed.

Reply:
> This doesn't seem right. It adds the key to setInternalKeyPool or setExternalKeyPool, but I don't think we want to do that?

To clarify, this wasn't just wasteful, but also a potentially dangerous bug right? LIke it could have caused new wallet addresses to be generated with the previous hd seed, and possibly stolen funds if money was sent to those addresses and the previous seed was unencrypted or shared with another wallet?